### ISSUE TITLE
Some configuration errors cause `scrcpy` client to hang

### ISSUE THREAD

**Issue by metayan** [Genymobile/scrcpy#5536]
## Environment

 - **OS:** macOS
 - **Scrcpy version:** 3.0/2.7/dev
 - **Installation method:** manual build (both client and server freshly built)
 - **Device model:** Sony
 - **Android version:** 9

Client hangs in a windowless state when trying
`lldb -- x/app/scrcpy --new-display=1920x1080`
or
`lldb -- x/app/scrcpy --video-source=camera`
on an Android 9 device.
Not even ctrl-c quits it.
Only KILL signal does.

Tried to fiddle with the `ConfigurationException` handling and to understand the termination procedure, but it's not clear enough to me to be able to fix this.

`lldb -- x/app/scrcpy --new-display=1920x1080`
```
(lldb) target create "x/app/scrcpy"
Current executable set to '/src/scm/android/scrcpy/x/app/scrcpy' (x86_64).
(lldb) settings set -- target.run-args  "--new-display=1920x1080"
(lldb) run
Process 9014 launched: '/src/scm/android/scrcpy/x/app/scrcpy' (x86_64)
scrcpy 3.0 <https://github.com/Genymobile/scrcpy>
INFO: ADB device found:
INFO:     -->   (usb)  xx                      device  xx
DEBUG: Device serial: xx
DEBUG: Using SCRCPY_SERVER_PATH: x/server/scrcpy-server
x/server/scrcpy-server: 1 file pushed, 0 skipped. 319.4 MB/s (90900 bytes in 0.000s)
[server] INFO: Device: [Sony] Sony xx (Android 9)
[server] ERROR: New virtual display is not supported before Android 10
DEBUG: Interrupting socket
DEBUG: Server disconnected
DEBUG: Server terminated
[hanged here with an empty client window open, so pressed ctrl-c]
Process 9014 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
    frame #0: 0x00007fff720d1746 libsystem_kernel.dylib`__semwait_signal + 10
libsystem_kernel.dylib`__semwait_signal:
->  0x7fff720d1746 <+10>: jae    0x7fff720d1750 ; <+20>
    0x7fff720d1748 <+12>: movq   %rax, %rdi
    0x7fff720d174b <+15>: jmp    0x7fff720d021d ; cerror
    0x7fff720d1750 <+20>: retq   
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
  * frame #0: 0x00007fff720d1746 libsystem_kernel.dylib`__semwait_signal + 10
    frame #1: 0x00007fff72054eea libsystem_c.dylib`nanosleep + 196
    frame #2: 0x0000000100902750 libSDL2-2.0.0.dylib`SDL_Delay_REAL + 83
    frame #3: 0x0000000100838352 libSDL2-2.0.0.dylib`SDL_WaitEventTimeout_REAL + 485
    frame #4: 0x000000010001e953 scrcpy`await_for_server(connected=0x00007ffeefbfef7e) at scrcpy.c:227:12
    frame #5: 0x000000010001d3ce scrcpy`scrcpy(options=0x00007ffeefbff0a8) at scrcpy.c:531:10
    frame #6: 0x0000000100003427 scrcpy`main_scrcpy(argc=2, argv=0x00007ffeefbff278) at main.c:92:52
    frame #7: 0x00000001000032f2 scrcpy`main(argc=2, argv=0x00007ffeefbff278) at main.c:111:12
    frame #8: 0x00007fff71f8dcc9 libdyld.dylib`start + 1
    frame #9: 0x00007fff71f8dcc9 libdyld.dylib`start + 1
(lldb) kill
Process 9014 exited with status = 9 (0x00000009) killed
```

Some `ConfigurationException`s exit nicely:
```
lldb -- x/app/scrcpy --video-encoder=junk
(lldb) target create "x/app/scrcpy"
Current executable set to '/src/scm/android/scrcpy/x/app/scrcpy' (x86_64).
(lldb) settings set -- target.run-args  "--video-encoder=junk"
(lldb) run
Process 9093 launched: '/src/scm/android/scrcpy/x/app/scrcpy' (x86_64)
scrcpy 3.0 <https://github.com/Genymobile/scrcpy>
INFO: ADB device found:
INFO:     -->   (usb)  xx                      device  xx
DEBUG: Device serial: xx
DEBUG: Using SCRCPY_SERVER_PATH: x/server/scrcpy-server
x/server/scrcpy-server: 1 file pushed, 0 skipped. 72.0 MB/s (90900 bytes in 0.001s)
[server] INFO: Device: [Sony] Sony xx (Android 9)
[server] DEBUG: Creating encoder by name: 'junk'
[server] WARN: Audio disabled: it is not supported before Android 11
[server] DEBUG: Audio encoder stopped
[server] ERROR: Video encoder 'junk' for h264 not found
List of video encoders:
    --video-codec=h264 --video-encoder=OMX.qcom.video.encoder.avc
    --video-codec=h264 --video-encoder=OMX.google.h264.encoder
    --video-codec=h265 --video-encoder=OMX.qcom.video.encoder.hevc
[server] DEBUG: Screen streaming stopped
[server] DEBUG: Device message sender stopped
[server] DEBUG: Controller stopped
DEBUG: Server disconnected
DEBUG: Server terminated
DEBUG: Server connected
DEBUG: Starting controller thread
DEBUG: Starting receiver thread
DEBUG: Receiver stopped
DEBUG: Using SCRCPY_ICON_PATH: app/data/icon.png
INFO: Renderer: metal
DEBUG: Trilinear filtering disabled (not an OpenGL renderer)
DEBUG: Demuxer 'video': starting thread
DEBUG: Demuxer 'audio': starting thread
WARN: Device disconnected
ERROR: Demuxer 'video': stream disabled due to connection error
WARN: Demuxer 'audio': stream explicitly disabled by the device
DEBUG: quit...
DEBUG: Controller stopped
Process 9093 exited with status = 2 (0x00000002) 
```
Even though it's a bit weird that the Server apparently connects after disconnecting and terminating.

https://github.com/Genymobile/scrcpy/blob/da8ade/app/src/scrcpy.c#L332-L333
looks worrying. Could it be that since no stream gets created, no "stream stopped" event will ever be received?

**Comment by rom1v** [comment]
Thank you for your report. I'll fix that.

**Comment by rom1v** [comment]
I cannot reproduce on Linux:

```
scrcpy 3.0 <https://github.com/Genymobile/scrcpy>
INFO: ADB device found:
INFO:     -->   (usb)  xx                device  Nexus_5
DEBUG: Device serial: xx
DEBUG: Using server: /usr/local/share/scrcpy/scrcpy-server
/usr/local/share/scrcpy/scrcpy-server: 1 file pushed, 0 skipped. 387.6 MB/s (90232 bytes in 0.000s)
[server] INFO: Device: [LGE] google Nexus 5 (Android 6.0.1)
[server] ERROR: New virtual display is not supported before Android 10
DEBUG: Interrupting socket
DEBUG: Server disconnected
DEBUG: Server terminated
ERROR: Server connection failed
```

Could you please add these logs:

```diff
diff --git app/src/scrcpy.c app/src/scrcpy.c
index dc9e237f3..f2b1ce464 100644
--- app/src/scrcpy.c
+++ app/src/scrcpy.c
@@ -312,6 +312,7 @@ sc_server_on_connection_failed(struct sc_server *server, void *userdata) {
     (void) server;
     (void) userdata;
 
+    LOGI("connection_failed");
     sc_push_event(SC_EVENT_SERVER_CONNECTION_FAILED);
 }
 
diff --git app/src/server.c app/src/server.c
index 584a3c347..87799c242 100644
--- app/src/server.c
+++ app/src/server.c
@@ -716,6 +716,7 @@ sc_server_on_terminated(void *userdata) {
     // then the client will be stuck forever on accept(). To avoid the problem,
     // wake up the accept() call (or any other) when the server dies, like on
     // stop() (it is safe to call interrupt() twice).
+    LOGI("sc_server_on_terminated()");
     sc_intr_interrupt(&server->intr);
 
     server->cbs->on_disconnected(server, server->cbs_userdata);
diff --git app/src/util/process.c app/src/util/process.c
index 9c4dcd9fc..4a6b14107 100644
--- app/src/util/process.c
+++ app/src/util/process.c
@@ -27,7 +27,9 @@ sc_pipe_read_all(sc_pipe pipe, char *data, size_t len) {
 static int
 run_observer(void *data) {
     struct sc_process_observer *observer = data;
+    LOGI("sc_process_observer wait");
     sc_process_wait(observer->pid, false); // ignore exit code
+    LOGI("sc_process_observer wait end");
 
     sc_mutex_lock(&observer->mutex);
     observer->terminated = true;
```

I get:

```
scrcpy 3.0 <https://github.com/Genymobile/scrcpy>
INFO: ADB device found:
INFO:     -->   (usb)  xx                device  Nexus_5
DEBUG: Device serial: xx
DEBUG: Using SCRCPY_SERVER_PATH: d/server/scrcpy-server
d/server/scrcpy-server: 1 file pushed, 0 skipped. 528.3 MB/s (159589 bytes in 0.000s)
INFO: sc_process_observer wait
[server] INFO: Device: [LGE] google Nexus 5 (Android 6.0.1)
[server] ERROR: New virtual display is not supported before Android 10
INFO: sc_process_observer wait end
INFO: sc_server_on_terminated()
DEBUG: Interrupting socket
DEBUG: Server disconnected
DEBUG: Server terminated
INFO: connection_failed
ERROR: Server connection failed
```

**Comment by metayan** [comment]
```
x/app/scrcpy --new-display
scrcpy 3.0 <https://github.com/Genymobile/scrcpy>
INFO: ADB device found:
INFO:     -->   (usb)  xx                      device  xx
DEBUG: Device serial: xx
DEBUG: Using SCRCPY_SERVER_PATH: x/server/scrcpy-server
x/server/scrcpy-server: 1 file pushed, 0 skipped. 93.0 MB/s (90877 bytes in 0.001s)
INFO: sc_process_observer wait
[server] INFO: Device: [Sony] Sony xx (Android 9)
[server] ERROR: New virtual display is not supported before Android 10
INFO: sc_process_observer wait end
INFO: sc_server_on_terminated()
DEBUG: Interrupting socket
DEBUG: Server disconnected
DEBUG: Server terminated
^CDEBUG: User requested to quit
^C^C^C^C^C
```
`killall -KILL scrcpy`
needed to terminate it.

When run in `lldb`, after the hang:
```
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
  * frame #0: 0x00007fff720d1746 libsystem_kernel.dylib`__semwait_signal + 10
    frame #1: 0x00007fff72054eea libsystem_c.dylib`nanosleep + 196
    frame #2: 0x0000000100902750 libSDL2-2.0.0.dylib`SDL_Delay_REAL + 83
    frame #3: 0x0000000100838352 libSDL2-2.0.0.dylib`SDL_WaitEventTimeout_REAL + 485
    frame #4: 0x000000010001e8c3 scrcpy`await_for_server(connected=0x00007ffeefbfef8e) at scrcpy.c:227:12
    frame #5: 0x000000010001d32e scrcpy`scrcpy(options=0x00007ffeefbff0b8) at scrcpy.c:532:10
    frame #6: 0x0000000100003387 scrcpy`main_scrcpy(argc=2, argv=0x00007ffeefbff280) at main.c:92:52
    frame #7: 0x0000000100003252 scrcpy`main(argc=2, argv=0x00007ffeefbff280) at main.c:111:12
    frame #8: 0x00007fff71f8dcc9 libdyld.dylib`start + 1
(lldb) kill
Process 21420 exited with status = 9 (0x00000009) killed
```

I see now that I have `sdl2` 2.30.8 installed and 2.30.9 is available. I can upgrade and check again.

**Comment by metayan** [comment]
> On don't pass on `connection_failed`.

I don't understand what you mean.

> ```
> thread apply all bt
> ```

My `lldb` doesn't understand this, but here are all the threads expanded in the GUI:
```
┌──<Threads>──────────────────────────────────────────┐
│ ◆─process 22068                                     │
│ ├─◆─thread #1: tid = 0x1d7a616, stop reason = signal│
│ │ ├─#0: __semwait_signal + 10                       │
│ │ ├─#1: nanosleep + 196                             │
│ │ ├─#2: SDL_Delay_REAL + 83                         │
│ │ ├─#3: SDL_WaitEventTimeout_REAL + 485             │
│ │ ├─#4: await_for_server + 35                       │
│ │ ├─#5: scrcpy + 1646                               │
│ │ ├─#6: main_scrcpy + 295                           │
│ │ ├─#7: main + 34                                   │
│ │ └─#8: start + 1                                   │
│ ├─◆─thread #2: tid = 0x1d7a647                      │
│ │ ├─#0: __workq_kernreturn + 10                     │
│ │ ├─#1: _pthread_wqthread + 390                     │
│ │ └─#2: start_wqthread + 15                         │
│ ├─◆─thread #3: tid = 0x1d7a648                      │
│ │ ├─#0: __accept + 10                               │
│ │ ├─#1: net_accept + 59                             │
│ │ ├─#2: net_accept_intr + 55                        │
│ │ ├─#3: sc_server_connect_to + 282                  │
│ │ ├─#4: run_server + 1721                           │
│ │ ├─#5: SDL_RunThread + 44                          │
│ │ ├─#6: RunThread + 9                               │
│ │ ├─#7: _pthread_start + 148                        │
│ │ └─#8: thread_start + 15                           │
│ ├─◆─thread #4: tid = 0x1d7a651                      │
│ │ ├─#0: __workq_kernreturn + 10                     │
│ │ ├─#1:                                             │
│ │ └─#2: start_wqthread + 15                         │
│ ├─◆─thread #5: tid = 0x1d7a674                      │
│ │ ├─#0: __workq_kernreturn + 10                     │
│ │ ├─#1: _pthread_wqthread + 390                     │
│ │ └─#2: start_wqthread + 15                         │
│ ├─◆─thread #7: tid = 0x1d7a6b1                      │
│ │ ├─#0: semaphore_wait_trap + 10                    │
│ │ ├─#1: caulk::mach::semaphore::wait() + 16         │
│ │ ├─#2: caulk::semaphore::timed_wait(double) + 106  │
│ │ ├─#3: caulk::concurrent::details::worker_thread::r│
│ │ ├─#4: void* caulk::thread_proxy<std::__1::tuple<ca│
│ │ ├─#5: _pthread_start + 148                        │
│ │ └─#6: thread_start + 15                           │
│ ├─◆─thread #8: tid = 0x1d7a6b2                      │
│ │ ├─#0: semaphore_wait_trap + 10                    │
│ │ ├─#1: caulk::mach::semaphore::wait() + 16         │
│ │ ├─#2: caulk::semaphore::timed_wait(double) + 106  │
│ │ ├─#3: caulk::concurrent::details::worker_thread::r│
│ │ ├─#4: void* caulk::thread_proxy<std::__1::tuple<ca│
│ │ ├─#5: _pthread_start + 148                        │
│ │ └─#6: thread_start + 15                           │
│ ├─◆─thread #9: tid = 0x1d7a6b5                      │
│ │ ├─#0: semaphore_timedwait_trap + 10               │
│ │ ├─#1: _dispatch_sema4_timedwait + 76              │
│ │ ├─#2: _dispatch_semaphore_wait_slow + 58          │
│ │ ├─#3: _dispatch_worker_thread + 284               │
│ │ ├─#4: _pthread_start + 148                        │
│ │ └─#5: thread_start + 15                           │
│ ├─◆─thread #10: tid = 0x1d7a6c2                     │
│ │ ├─#0: mach_msg_trap + 10                          │
│ │ ├─#1: mach_msg + 60                               │
│ │ ├─#2: __CFRunLoopServiceMachPort + 247            │
│ │ ├─#3: __CFRunLoopRun + 1319                       │
│ │ ├─#4: CFRunLoopRunSpecific + 462                  │
│ │ ├─#5: _NSEventThread + 132                        │
│ │ ├─#6: _pthread_start + 148                        │
│ │ └─#7: thread_start + 15                           │
│ ├─◆─thread #11: tid = 0x1d7a6c3                     │
│ │ ├─#0: __workq_kernreturn + 10                     │
│ │ ├─#1: _pthread_wqthread + 390                     │
│ │ └─#2: start_wqthread + 15                         │
│ ├─◆─thread #12: tid = 0x1d7a6c4                     │
│ │ ├─#0: __workq_kernreturn + 10                     │
│ │ ├─#1: _pthread_wqthread + 390                     │
│ │ └─#2: start_wqthread + 15                         │
│ ├─◆─thread #13: tid = 0x1d7a6c5                     │
│ │ ├─#0: __workq_kernreturn + 10                     │
│ │ ├─#1: _pthread_wqthread + 390                     │
│ │ └─#2: start_wqthread + 15                         │
│ ├─◆─thread #14: tid = 0x1d7a6c6                     │
│ │ ├─#0: __workq_kernreturn + 10                     │
│ │ ├─#1: _pthread_wqthread + 390                     │
│ │ └─#2: start_wqthread + 15                         │
│ └─◆─thread #15: tid = 0x1d7a6c7                     │
│   ├─#0: __workq_kernreturn + 10                     │
│   ├─#1: _pthread_wqthread + 390                     │
│   └─#2: start_wqthread + 15                         │
│                                                     │
```

**Comment by rom1v** [comment]
> My lldb doesn't understand this

ChatGPT says:

> In LLDB, the equivalent of thread apply all bt from GDB is the command:
>
>    thread backtrace all
>
> Or its shorthand:
>
>    thread bt all
>
> This command displays the backtrace for all threads, similar to how thread apply all bt works in GDB.

**Comment by metayan** [comment]
Thanks.
```thread backtrace all```
worked. (Not the shorthand, though.)

Anyway:
```
(lldb) thread backtrace all
  thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
    frame #0: 0x00007fff720d1746 libsystem_kernel.dylib`__semwait_signal + 10
    frame #1: 0x00007fff72054eea libsystem_c.dylib`nanosleep + 196
    frame #2: 0x0000000100902750 libSDL2-2.0.0.dylib`SDL_Delay_REAL + 83
    frame #3: 0x0000000100838352 libSDL2-2.0.0.dylib`SDL_WaitEventTimeout_REAL + 485
    frame #4: 0x000000010001e8c3 scrcpy`await_for_server(connected=0x00007ffeefbfef8e) at scrcpy.c:227:12
    frame #5: 0x000000010001d32e scrcpy`scrcpy(options=0x00007ffeefbff0b8) at scrcpy.c:532:10
    frame #6: 0x0000000100003387 scrcpy`main_scrcpy(argc=2, argv=0x00007ffeefbff280) at main.c:92:52
    frame #7: 0x0000000100003252 scrcpy`main(argc=2, argv=0x00007ffeefbff280) at main.c:111:12
    frame #8: 0x00007fff71f8dcc9 libdyld.dylib`start + 1
* thread #2
  * frame #0: 0x00007fff720d04be libsystem_kernel.dylib`__workq_kernreturn + 10
    frame #1: 0x00007fff7218eaa1 libsystem_pthread.dylib`_pthread_wqthread + 390
    frame #2: 0x00007fff7218db77 libsystem_pthread.dylib`start_wqthread + 15
  thread #3, name = 'scrcpy-server'
    frame #0: 0x00007fff720d53ae libsystem_kernel.dylib`__accept + 10
    frame #1: 0x00000001000292fb scrcpy`net_accept(server_socket=3) at net.c:202:9
    frame #2: 0x0000000100029807 scrcpy`net_accept_intr(intr=0x00000001000c4888, server_socket=3) at net_intr.c:38:24
    frame #3: 0x000000010002405a scrcpy`sc_server_connect_to(server=0x00000001000c4740, info=0x00000001000c4828) at server.c:574:17
    frame #4: 0x0000000100022099 scrcpy`run_server(data=0x00000001000c4740) at server.c:1046:10
    frame #5: 0x000000010087c7d8 libSDL2-2.0.0.dylib`SDL_RunThread + 44
    frame #6: 0x00000001008e733e libSDL2-2.0.0.dylib`RunThread + 9
    frame #7: 0x00007fff72192109 libsystem_pthread.dylib`_pthread_start + 148
    frame #8: 0x00007fff7218db8b libsystem_pthread.dylib`thread_start + 15
  thread #4
    frame #0: 0x00007fff7218db68 libsystem_pthread.dylib`start_wqthread
  thread #5
    frame #0: 0x00007fff7218db68 libsystem_pthread.dylib`start_wqthread
  thread #7, name = 'AMCP Logging Spool'
    frame #0: 0x00007fff720cee26 libsystem_kernel.dylib`semaphore_wait_trap + 10
    frame #1: 0x00007fff6bb5fb16 caulk`caulk::mach::semaphore::wait() + 16
    frame #2: 0x00007fff6bb5f9b2 caulk`caulk::semaphore::timed_wait(double) + 106
    frame #3: 0x00007fff6bb5f7c4 caulk`caulk::concurrent::details::worker_thread::run() + 30
    frame #4: 0x00007fff6bb5f1e4 caulk`void* caulk::thread_proxy<std::__1::tuple<caulk::thread::attributes, void (caulk::concurrent::details::worker_thread::*)(), std::__1::tuple<caulk::concurrent::details::worker_thread*>>>(void*) + 45
    frame #5: 0x00007fff72192109 libsystem_pthread.dylib`_pthread_start + 148
    frame #6: 0x00007fff7218db8b libsystem_pthread.dylib`thread_start + 15
  thread #8
    frame #0: 0x00007fff720cee26 libsystem_kernel.dylib`semaphore_wait_trap + 10
    frame #1: 0x00007fff6bb5fb16 caulk`caulk::mach::semaphore::wait() + 16
    frame #2: 0x00007fff6bb5f9b2 caulk`caulk::semaphore::timed_wait(double) + 106
    frame #3: 0x00007fff6bb5f7c4 caulk`caulk::concurrent::details::worker_thread::run() + 30
    frame #4: 0x00007fff6bb5f1e4 caulk`void* caulk::thread_proxy<std::__1::tuple<caulk::thread::attributes, void (caulk::concurrent::details::worker_thread::*)(), std::__1::tuple<caulk::concurrent::details::worker_thread*>>>(void*) + 45
    frame #5: 0x00007fff72192109 libsystem_pthread.dylib`_pthread_start + 148
    frame #6: 0x00007fff7218db8b libsystem_pthread.dylib`thread_start + 15
  thread #9
    frame #0: 0x00007fff720cee3e libsystem_kernel.dylib`semaphore_timedwait_trap + 10
    frame #1: 0x00007fff71f34b6b libdispatch.dylib`_dispatch_sema4_timedwait + 76
    frame #2: 0x00007fff71f34f97 libdispatch.dylib`_dispatch_semaphore_wait_slow + 58
    frame #3: 0x00007fff71f42795 libdispatch.dylib`_dispatch_worker_thread + 284
    frame #4: 0x00007fff72192109 libsystem_pthread.dylib`_pthread_start + 148
    frame #5: 0x00007fff7218db8b libsystem_pthread.dylib`thread_start + 15
  thread #10, name = 'com.apple.NSEventThread'
    frame #0: 0x00007fff720cedea libsystem_kernel.dylib`mach_msg_trap + 10
    frame #1: 0x00007fff720cf160 libsystem_kernel.dylib`mach_msg + 60
    frame #2: 0x00007fff38309135 CoreFoundation`__CFRunLoopServiceMachPort + 247
    frame #3: 0x00007fff38307c02 CoreFoundation`__CFRunLoopRun + 1319
    frame #4: 0x00007fff3830707e CoreFoundation`CFRunLoopRunSpecific + 462
    frame #5: 0x00007fff3571dbf4 AppKit`_NSEventThread + 132
    frame #6: 0x00007fff72192109 libsystem_pthread.dylib`_pthread_start + 148
    frame #7: 0x00007fff7218db8b libsystem_pthread.dylib`thread_start + 15
  thread #11
    frame #0: 0x00007fff720d04be libsystem_kernel.dylib`__workq_kernreturn + 10
    frame #1: 0x00007fff7218eaa1 libsystem_pthread.dylib`_pthread_wqthread + 390
    frame #2: 0x00007fff7218db77 libsystem_pthread.dylib`start_wqthread + 15
  thread #12
    frame #0: 0x00007fff720d04be libsystem_kernel.dylib`__workq_kernreturn + 10
    frame #1: 0x00007fff7218eaa1 libsystem_pthread.dylib`_pthread_wqthread + 390
    frame #2: 0x00007fff7218db77 libsystem_pthread.dylib`start_wqthread + 15
  thread #13
    frame #0: 0x00007fff720d04be libsystem_kernel.dylib`__workq_kernreturn + 10
    frame #1: 0x00007fff7218eaa1 libsystem_pthread.dylib`_pthread_wqthread + 390
    frame #2: 0x00007fff7218db77 libsystem_pthread.dylib`start_wqthread + 15
  thread #14
    frame #0: 0x00007fff7218db68 libsystem_pthread.dylib`start_wqthread
  thread #15
    frame #0: 0x00007fff7218db68 libsystem_pthread.dylib`start_wqthread
```

**Comment by rom1v** [comment]
> ```
> │ ├─◆─thread #3: tid = 0x1d7a648                      │
> │ │ ├─#0: __accept + 10                               │
> │ │ ├─#1: net_accept + 59                             │
> │ │ ├─#2: net_accept_intr + 55                        │
> ```

So it seems you are blocked on `net_accept`, but your logs show:

> DEBUG: Interrupting socket

So it seems `shutdown()` does not interrupt an `accept()`…

Could you please add these logs:

```diff
diff --git app/src/util/net.c app/src/util/net.c
index d43d1c7aa..81500cc79 100644
--- app/src/util/net.c
+++ app/src/util/net.c
@@ -195,8 +195,10 @@ net_accept(sc_socket server_socket) {
     socklen_t sinsize = sizeof(csin);
 
 #ifdef HAVE_SOCK_CLOEXEC
+    LOGI("net_accept()");
     sc_raw_socket raw_sock =
         accept4(raw_server_socket, (SOCKADDR *) &csin, &sinsize, SOCK_CLOEXEC);
+    LOGI("net_accept() end");
 #else
     sc_raw_socket raw_sock =
         accept(raw_server_socket, (SOCKADDR *) &csin, &sinsize);
@@ -254,6 +256,7 @@ net_interrupt(sc_socket socket) {
     }
     return true;
 #else
+    LOGI("shutdown socket");
     return !shutdown(raw_sock, SHUT_RDWR);
 #endif
 }
```

I get:

```console
$ ./run d --new-display
scrcpy 3.0 <https://github.com/Genymobile/scrcpy>
INFO: ADB device found:
INFO:     -->   (usb)  xx                device  Nexus_5
DEBUG: Device serial: xx
DEBUG: Using SCRCPY_SERVER_PATH: d/server/scrcpy-server
d/server/scrcpy-server: 1 file pushed, 0 skipped. 743.9 MB/s (159589 bytes in 0.000s)
INFO: net_accept()
INFO: sc_process_observer wait
[server] INFO: Device: [LGE] google Nexus 5 (Android 6.0.1)
[server] ERROR: New virtual display is not supported before Android 10
INFO: sc_process_observer wait end
INFO: sc_server_on_terminated()
DEBUG: Interrupting socket
INFO: shutdown socket
DEBUG: Server disconnected
DEBUG: Server terminated
INFO: net_accept() end
INFO: connection_failed
ERROR: Server connection failed
```

**Comment by metayan** [comment]
With all the extra logging:

```lldb -- x/app/scrcpy --new-display```
```
(lldb) target create "x/app/scrcpy"
Current executable set to '/src/scm/android/scrcpy/x/app/scrcpy' (x86_64).
(lldb) settings set -- target.run-args  "--new-display"
(lldb) run
Process 22552 launched: '/src/scm/android/scrcpy/x/app/scrcpy' (x86_64)
scrcpy 3.0 <https://github.com/Genymobile/scrcpy>
INFO: ADB device found:
INFO:     -->   (usb)  xx                      device  G8441
DEBUG: Device serial: xx
DEBUG: Using SCRCPY_SERVER_PATH: x/server/scrcpy-server
x/server/scrcpy-server: 1 file pushed, 0 skipped. 259.9 MB/s (90877 bytes in 0.000s)
INFO: sc_process_observer wait
[server] INFO: Device: [Sony] Sony G8441 (Android 9)
[server] ERROR: New virtual display is not supported before Android 10
INFO: sc_process_observer wait end
INFO: sc_server_on_terminated()
DEBUG: Interrupting socket
INFO: shutdown socket
DEBUG: Server disconnected
DEBUG: Server terminated
Process 22552 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
    frame #0: 0x00007fff720d1746 libsystem_kernel.dylib`__semwait_signal + 10
libsystem_kernel.dylib`__semwait_signal:
->  0x7fff720d1746 <+10>: jae    0x7fff720d1750 ; <+20>
    0x7fff720d1748 <+12>: movq   %rax, %rdi
    0x7fff720d174b <+15>: jmp    0x7fff720d021d ; cerror
    0x7fff720d1750 <+20>: retq   

(lldb) thread backtrace all
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
  * frame #0: 0x00007fff720d1746 libsystem_kernel.dylib`__semwait_signal + 10
    frame #1: 0x00007fff72054eea libsystem_c.dylib`nanosleep + 196
    frame #2: 0x0000000100902750 libSDL2-2.0.0.dylib`SDL_Delay_REAL + 83
    frame #3: 0x0000000100838352 libSDL2-2.0.0.dylib`SDL_WaitEventTimeout_REAL + 485
    frame #4: 0x000000010001e8a3 scrcpy`await_for_server(connected=0x00007ffeefbfef8e) at scrcpy.c:227:12
    frame #5: 0x000000010001d30e scrcpy`scrcpy(options=0x00007ffeefbff0b8) at scrcpy.c:532:10
    frame #6: 0x0000000100003367 scrcpy`main_scrcpy(argc=2, argv=0x00007ffeefbff280) at main.c:92:52
    frame #7: 0x0000000100003232 scrcpy`main(argc=2, argv=0x00007ffeefbff280) at main.c:111:12
    frame #8: 0x00007fff71f8dcc9 libdyld.dylib`start + 1
  thread #3, name = 'scrcpy-server'
    frame #0: 0x00007fff720d53ae libsystem_kernel.dylib`__accept + 10
    frame #1: 0x00000001000292db scrcpy`net_accept(server_socket=3) at net.c:204:9
    frame #2: 0x00000001000297f7 scrcpy`net_accept_intr(intr=0x00000001000c4888, server_socket=3) at net_intr.c:38:24
    frame #3: 0x000000010002403a scrcpy`sc_server_connect_to(server=0x00000001000c4740, info=0x00000001000c4828) at server.c:574:17
    frame #4: 0x0000000100022079 scrcpy`run_server(data=0x00000001000c4740) at server.c:1046:10
    frame #5: 0x000000010087c7d8 libSDL2-2.0.0.dylib`SDL_RunThread + 44
    frame #6: 0x00000001008e733e libSDL2-2.0.0.dylib`RunThread + 9
    frame #7: 0x00007fff72192109 libsystem_pthread.dylib`_pthread_start + 148
    frame #8: 0x00007fff7218db8b libsystem_pthread.dylib`thread_start + 15
  thread #7, name = 'AMCP Logging Spool'
    frame #0: 0x00007fff720cee26 libsystem_kernel.dylib`semaphore_wait_trap + 10
    frame #1: 0x00007fff6bb5fb16 caulk`caulk::mach::semaphore::wait() + 16
    frame #2: 0x00007fff6bb5f9b2 caulk`caulk::semaphore::timed_wait(double) + 106
    frame #3: 0x00007fff6bb5f7c4 caulk`caulk::concurrent::details::worker_thread::run() + 30
    frame #4: 0x00007fff6bb5f1e4 caulk`void* caulk::thread_proxy<std::__1::tuple<caulk::thread::attributes, void (caulk::concurrent::details::worker_thread::*)(), std::__1::tuple<caulk::concurrent::details::worker_thread*>>>(void*) + 45
    frame #5: 0x00007fff72192109 libsystem_pthread.dylib`_pthread_start + 148
    frame #6: 0x00007fff7218db8b libsystem_pthread.dylib`thread_start + 15
  thread #8
    frame #0: 0x00007fff720cee26 libsystem_kernel.dylib`semaphore_wait_trap + 10
    frame #1: 0x00007fff6bb5fb16 caulk`caulk::mach::semaphore::wait() + 16
    frame #2: 0x00007fff6bb5f9b2 caulk`caulk::semaphore::timed_wait(double) + 106
    frame #3: 0x00007fff6bb5f7c4 caulk`caulk::concurrent::details::worker_thread::run() + 30
    frame #4: 0x00007fff6bb5f1e4 caulk`void* caulk::thread_proxy<std::__1::tuple<caulk::thread::attributes, void (caulk::concurrent::details::worker_thread::*)(), std::__1::tuple<caulk::concurrent::details::worker_thread*>>>(void*) + 45
    frame #5: 0x00007fff72192109 libsystem_pthread.dylib`_pthread_start + 148
    frame #6: 0x00007fff7218db8b libsystem_pthread.dylib`thread_start + 15
  thread #10, name = 'com.apple.NSEventThread'
    frame #0: 0x00007fff720cedea libsystem_kernel.dylib`mach_msg_trap + 10
    frame #1: 0x00007fff720cf160 libsystem_kernel.dylib`mach_msg + 60
    frame #2: 0x00007fff38309135 CoreFoundation`__CFRunLoopServiceMachPort + 247
    frame #3: 0x00007fff38307c02 CoreFoundation`__CFRunLoopRun + 1319
    frame #4: 0x00007fff3830707e CoreFoundation`CFRunLoopRunSpecific + 462
    frame #5: 0x00007fff3571dbf4 AppKit`_NSEventThread + 132
    frame #6: 0x00007fff72192109 libsystem_pthread.dylib`_pthread_start + 148
    frame #7: 0x00007fff7218db8b libsystem_pthread.dylib`thread_start + 15
  thread #11
    frame #0: 0x00007fff7218db68 libsystem_pthread.dylib`start_wqthread
(lldb) kill
Process 22552 exited with status = 9 (0x00000009) killed
```

**Comment by rom1v** [comment]
Hmmm… more:

```diff
diff --git app/src/server.c app/src/server.c
index 0e449deae..cf9e82486 100644
--- app/src/server.c
+++ app/src/server.c
@@ -1044,12 +1044,18 @@ run_server(void *data) {
         goto error_connection_failed;
     }
 
+    LOGI("sc_server_connect_to()");
     ok = sc_server_connect_to(server, &server->info);
+    LOGI("sc_server_connect_to() returned %s", ok ? "true" : "false");
     // The tunnel is always closed by server_connect_to()
     if (!ok) {
+        LOGI("sc_process_terminate(%d)", pid);
         sc_process_terminate(pid);
+        LOGI("sc_process_wait(%d, true)", pid);
         sc_process_wait(pid, true); // ignore exit code
+        LOGI("sc_process_observer_join()");
         sc_process_observer_join(&observer);
+        LOGI("sc_process_observer_destroy()");
         sc_process_observer_destroy(&observer);
         goto error_connection_failed;
     }
```

**Comment by metayan** [comment]
```lldb -- x/app/scrcpy --new-display```
```
(lldb) target create "x/app/scrcpy"
Current executable set to '/src/scm/android/scrcpy/x/app/scrcpy' (x86_64).
(lldb) settings set -- target.run-args  "--new-display"
(lldb) run
Process 22996 launched: '/src/scm/android/scrcpy/x/app/scrcpy' (x86_64)
scrcpy 3.0 <https://github.com/Genymobile/scrcpy>
INFO: ADB device found:
INFO:     -->   (usb)  xx                      device  G8441
DEBUG: Device serial: xx
DEBUG: Using SCRCPY_SERVER_PATH: x/server/scrcpy-server
x/server/scrcpy-server: 1 file pushed, 0 skipped. 301.2 MB/s (90877 bytes in 0.000s)
INFO: sc_server_connect_to()
INFO: sc_process_observer wait
[server] INFO: Device: [Sony] Sony G8441 (Android 9)
[server] ERROR: New virtual display is not supported before Android 10
INFO: sc_process_observer wait end
INFO: sc_server_on_terminated()
DEBUG: Interrupting socket
INFO: shutdown socket
DEBUG: Server disconnected
DEBUG: Server terminated
[----- hanged here -----]
Process 22996 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
    frame #0: 0x00007fff720d1746 libsystem_kernel.dylib`__semwait_signal + 10
libsystem_kernel.dylib`__semwait_signal:
->  0x7fff720d1746 <+10>: jae    0x7fff720d1750 ; <+20>
    0x7fff720d1748 <+12>: movq   %rax, %rdi
    0x7fff720d174b <+15>: jmp    0x7fff720d021d ; cerror
    0x7fff720d1750 <+20>: retq   
(lldb) thread backtrace all
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
  * frame #0: 0x00007fff720d1746 libsystem_kernel.dylib`__semwait_signal + 10
    frame #1: 0x00007fff72054eea libsystem_c.dylib`nanosleep + 196
    frame #2: 0x0000000100902750 libSDL2-2.0.0.dylib`SDL_Delay_REAL + 83
    frame #3: 0x0000000100838352 libSDL2-2.0.0.dylib`SDL_WaitEventTimeout_REAL + 485
    frame #4: 0x000000010001e783 scrcpy`await_for_server(connected=0x00007ffeefbfef8e) at scrcpy.c:227:12
    frame #5: 0x000000010001d1ee scrcpy`scrcpy(options=0x00007ffeefbff0b8) at scrcpy.c:532:10
    frame #6: 0x0000000100003247 scrcpy`main_scrcpy(argc=2, argv=0x00007ffeefbff280) at main.c:92:52
    frame #7: 0x0000000100003112 scrcpy`main(argc=2, argv=0x00007ffeefbff280) at main.c:111:12
    frame #8: 0x00007fff71f8dcc9 libdyld.dylib`start + 1
  thread #3, name = 'scrcpy-server'
    frame #0: 0x00007fff720d53ae libsystem_kernel.dylib`__accept + 10
    frame #1: 0x000000010002923b scrcpy`net_accept(server_socket=4) at net.c:204:9
    frame #2: 0x0000000100029757 scrcpy`net_accept_intr(intr=0x00000001000c4888, server_socket=4) at net_intr.c:38:24
    frame #3: 0x0000000100023f9a scrcpy`sc_server_connect_to(server=0x00000001000c4740, info=0x00000001000c4828) at server.c:574:17
    frame #4: 0x0000000100021f69 scrcpy`run_server(data=0x00000001000c4740) at server.c:1047:10
    frame #5: 0x000000010087c7d8 libSDL2-2.0.0.dylib`SDL_RunThread + 44
    frame #6: 0x00000001008e733e libSDL2-2.0.0.dylib`RunThread + 9
    frame #7: 0x00007fff72192109 libsystem_pthread.dylib`_pthread_start + 148
    frame #8: 0x00007fff7218db8b libsystem_pthread.dylib`thread_start + 15
  thread #6, name = 'AMCP Logging Spool'
    frame #0: 0x00007fff720cee26 libsystem_kernel.dylib`semaphore_wait_trap + 10
    frame #1: 0x00007fff6bb5fb16 caulk`caulk::mach::semaphore::wait() + 16
    frame #2: 0x00007fff6bb5f9b2 caulk`caulk::semaphore::timed_wait(double) + 106
    frame #3: 0x00007fff6bb5f7c4 caulk`caulk::concurrent::details::worker_thread::run() + 30
    frame #4: 0x00007fff6bb5f1e4 caulk`void* caulk::thread_proxy<std::__1::tuple<caulk::thread::attributes, void (caulk::concurrent::details::worker_thread::*)(), std::__1::tuple<caulk::concurrent::details::worker_thread*>>>(void*) + 45
    frame #5: 0x00007fff72192109 libsystem_pthread.dylib`_pthread_start + 148
    frame #6: 0x00007fff7218db8b libsystem_pthread.dylib`thread_start + 15
  thread #7
    frame #0: 0x00007fff720cee26 libsystem_kernel.dylib`semaphore_wait_trap + 10
    frame #1: 0x00007fff6bb5fb16 caulk`caulk::mach::semaphore::wait() + 16
    frame #2: 0x00007fff6bb5f9b2 caulk`caulk::semaphore::timed_wait(double) + 106
    frame #3: 0x00007fff6bb5f7c4 caulk`caulk::concurrent::details::worker_thread::run() + 30
    frame #4: 0x00007fff6bb5f1e4 caulk`void* caulk::thread_proxy<std::__1::tuple<caulk::thread::attributes, void (caulk::concurrent::details::worker_thread::*)(), std::__1::tuple<caulk::concurrent::details::worker_thread*>>>(void*) + 45
    frame #5: 0x00007fff72192109 libsystem_pthread.dylib`_pthread_start + 148
    frame #6: 0x00007fff7218db8b libsystem_pthread.dylib`thread_start + 15
  thread #10, name = 'com.apple.NSEventThread'
    frame #0: 0x00007fff720cedea libsystem_kernel.dylib`mach_msg_trap + 10
    frame #1: 0x00007fff720cf160 libsystem_kernel.dylib`mach_msg + 60
    frame #2: 0x00007fff38309135 CoreFoundation`__CFRunLoopServiceMachPort + 247
    frame #3: 0x00007fff38307c02 CoreFoundation`__CFRunLoopRun + 1319
    frame #4: 0x00007fff3830707e CoreFoundation`CFRunLoopRunSpecific + 462
    frame #5: 0x00007fff3571dbf4 AppKit`_NSEventThread + 132
    frame #6: 0x00007fff72192109 libsystem_pthread.dylib`_pthread_start + 148
    frame #7: 0x00007fff7218db8b libsystem_pthread.dylib`thread_start + 15
  thread #11
    frame #0: 0x00007fff7218db68 libsystem_pthread.dylib`start_wqthread
  thread #12
    frame #0: 0x00007fff7218db68 libsystem_pthread.dylib`start_wqthread
  thread #13
    frame #0: 0x00007fff7218db68 libsystem_pthread.dylib`start_wqthread
(lldb) kill
Process 22996 exited with status = 9 (0x00000009) killed
```

**Comment by metayan** [comment]
```lldb -- x/app/scrcpy --new-display```
```
(lldb) target create "x/app/scrcpy"
Current executable set to '/src/scm/android/scrcpy/x/app/scrcpy' (x86_64).
(lldb) settings set -- target.run-args  "--new-display"
(lldb) run
Process 23450 launched: '/src/scm/android/scrcpy/x/app/scrcpy' (x86_64)
scrcpy 3.0 <https://github.com/Genymobile/scrcpy>
INFO: ADB device found:
INFO:     -->   (usb)  xx                      device  G8441
DEBUG: Device serial: xx
DEBUG: Using SCRCPY_SERVER_PATH: x/server/scrcpy-server
x/server/scrcpy-server: 1 file pushed, 0 skipped. 75.1 MB/s (90877 bytes in 0.001s)
INFO: sc_server_connect_to()
INFO: accept()
INFO: sc_process_observer wait
[server] INFO: Device: [Sony] Sony G8441 (Android 9)
[server] ERROR: New virtual display is not supported before Android 10
INFO: sc_process_observer wait end
INFO: sc_server_on_terminated()
DEBUG: Interrupting socket
INFO: shutdown socket
DEBUG: Server disconnected
DEBUG: Server terminated
[----- hanged here -----]
Process 23450 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
    frame #0: 0x00007fff720d1746 libsystem_kernel.dylib`__semwait_signal + 10
libsystem_kernel.dylib`__semwait_signal:
->  0x7fff720d1746 <+10>: jae    0x7fff720d1750 ; <+20>
    0x7fff720d1748 <+12>: movq   %rax, %rdi
    0x7fff720d174b <+15>: jmp    0x7fff720d021d ; cerror
    0x7fff720d1750 <+20>: retq   
(lldb) thread backtrace all
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
  * frame #0: 0x00007fff720d1746 libsystem_kernel.dylib`__semwait_signal + 10
    frame #1: 0x00007fff72054eea libsystem_c.dylib`nanosleep + 196
    frame #2: 0x0000000100902750 libSDL2-2.0.0.dylib`SDL_Delay_REAL + 83
    frame #3: 0x0000000100838352 libSDL2-2.0.0.dylib`SDL_WaitEventTimeout_REAL + 485
    frame #4: 0x000000010001e753 scrcpy`await_for_server(connected=0x00007ffeefbfef8e) at scrcpy.c:227:12
    frame #5: 0x000000010001d1be scrcpy`scrcpy(options=0x00007ffeefbff0b8) at scrcpy.c:532:10
    frame #6: 0x0000000100003217 scrcpy`main_scrcpy(argc=2, argv=0x00007ffeefbff280) at main.c:92:52
    frame #7: 0x00000001000030e2 scrcpy`main(argc=2, argv=0x00007ffeefbff280) at main.c:111:12
    frame #8: 0x00007fff71f8dcc9 libdyld.dylib`start + 1
  thread #3, name = 'scrcpy-server'
    frame #0: 0x00007fff720d53ae libsystem_kernel.dylib`__accept + 10
    frame #1: 0x000000010002921b scrcpy`net_accept(server_socket=3) at net.c:205:9
    frame #2: 0x0000000100029747 scrcpy`net_accept_intr(intr=0x00000001000c4888, server_socket=3) at net_intr.c:38:24
    frame #3: 0x0000000100023f6a scrcpy`sc_server_connect_to(server=0x00000001000c4740, info=0x00000001000c4828) at server.c:574:17
    frame #4: 0x0000000100021f39 scrcpy`run_server(data=0x00000001000c4740) at server.c:1047:10
    frame #5: 0x000000010087c7d8 libSDL2-2.0.0.dylib`SDL_RunThread + 44
    frame #6: 0x00000001008e733e libSDL2-2.0.0.dylib`RunThread + 9
    frame #7: 0x00007fff72192109 libsystem_pthread.dylib`_pthread_start + 148
    frame #8: 0x00007fff7218db8b libsystem_pthread.dylib`thread_start + 15
  thread #6, name = 'AMCP Logging Spool'
    frame #0: 0x00007fff720cee26 libsystem_kernel.dylib`semaphore_wait_trap + 10
    frame #1: 0x00007fff6bb5fb16 caulk`caulk::mach::semaphore::wait() + 16
    frame #2: 0x00007fff6bb5f9b2 caulk`caulk::semaphore::timed_wait(double) + 106
    frame #3: 0x00007fff6bb5f7c4 caulk`caulk::concurrent::details::worker_thread::run() + 30
    frame #4: 0x00007fff6bb5f1e4 caulk`void* caulk::thread_proxy<std::__1::tuple<caulk::thread::attributes, void (caulk::concurrent::details::worker_thread::*)(), std::__1::tuple<caulk::concurrent::details::worker_thread*>>>(void*) + 45
    frame #5: 0x00007fff72192109 libsystem_pthread.dylib`_pthread_start + 148
    frame #6: 0x00007fff7218db8b libsystem_pthread.dylib`thread_start + 15
  thread #7
    frame #0: 0x00007fff720cee26 libsystem_kernel.dylib`semaphore_wait_trap + 10
    frame #1: 0x00007fff6bb5fb16 caulk`caulk::mach::semaphore::wait() + 16
    frame #2: 0x00007fff6bb5f9b2 caulk`caulk::semaphore::timed_wait(double) + 106
    frame #3: 0x00007fff6bb5f7c4 caulk`caulk::concurrent::details::worker_thread::run() + 30
    frame #4: 0x00007fff6bb5f1e4 caulk`void* caulk::thread_proxy<std::__1::tuple<caulk::thread::attributes, void (caulk::concurrent::details::worker_thread::*)(), std::__1::tuple<caulk::concurrent::details::worker_thread*>>>(void*) + 45
    frame #5: 0x00007fff72192109 libsystem_pthread.dylib`_pthread_start + 148
    frame #6: 0x00007fff7218db8b libsystem_pthread.dylib`thread_start + 15
  thread #9, name = 'com.apple.NSEventThread'
    frame #0: 0x00007fff720cedea libsystem_kernel.dylib`mach_msg_trap + 10
    frame #1: 0x00007fff720cf160 libsystem_kernel.dylib`mach_msg + 60
    frame #2: 0x00007fff38309135 CoreFoundation`__CFRunLoopServiceMachPort + 247
    frame #3: 0x00007fff38307c02 CoreFoundation`__CFRunLoopRun + 1319
    frame #4: 0x00007fff3830707e CoreFoundation`CFRunLoopRunSpecific + 462
    frame #5: 0x00007fff3571dbf4 AppKit`_NSEventThread + 132
    frame #6: 0x00007fff72192109 libsystem_pthread.dylib`_pthread_start + 148
    frame #7: 0x00007fff7218db8b libsystem_pthread.dylib`thread_start + 15
  thread #10
    frame #0: 0x00007fff7218db68 libsystem_pthread.dylib`start_wqthread
  thread #11
    frame #0: 0x0000000000000000
(lldb) kill
Process 23450 exited with status = 9 (0x00000009) killed
```

**Comment by metayan** [comment]
Just for reference with all the logging, when it exits nicely:

```lldb -- x/app/scrcpy --video-encoder=junk```

```
(lldb) target create "x/app/scrcpy"
Current executable set to '/src/scm/android/scrcpy/x/app/scrcpy' (x86_64).
(lldb) settings set -- target.run-args  "--video-encoder=junk"
(lldb) run
Process 23503 launched: '/src/scm/android/scrcpy/x/app/scrcpy' (x86_64)
scrcpy 3.0 <https://github.com/Genymobile/scrcpy>
INFO: ADB device found:
INFO:     -->   (usb)  xx                      device  G8441
DEBUG: Device serial: xx
DEBUG: Using SCRCPY_SERVER_PATH: x/server/scrcpy-server
x/server/scrcpy-server: 1 file pushed, 0 skipped. 88.0 MB/s (90877 bytes in 0.001s)
INFO: sc_process_observer wait
INFO: sc_server_connect_to()
INFO: accept()
[server] INFO: Device: [Sony] Sony G8441 (Android 9)
INFO: accept() end
INFO: accept()
INFO: accept() end
INFO: accept()
INFO: accept() end
[server] DEBUG: Creating encoder by name: 'junk'
[server] WARN: Audio disabled: it is not supported before Android 11
[server] DEBUG: Audio encoder stopped
INFO: sc_server_connect_to() returned true
[server] DEBUG: Screen streaming stopped
[server] ERROR: Video encoder 'junk' for h264 not found
List of video encoders:
    --video-codec=h264 --video-encoder=OMX.qcom.video.encoder.avc
    --video-codec=h264 --video-encoder=OMX.google.h264.encoder
    --video-codec=h265 --video-encoder=OMX.qcom.video.encoder.hevc
[server] DEBUG: Device message sender stopped
[server] DEBUG: Controller stopped
INFO: sc_process_observer wait end
INFO: sc_server_on_terminated()
DEBUG: Server disconnected
DEBUG: Server terminated
DEBUG: Server connected
DEBUG: Starting controller thread
DEBUG: Starting receiver thread
DEBUG: Receiver stopped
DEBUG: Using SCRCPY_ICON_PATH: app/data/icon.png
INFO: Renderer: metal
DEBUG: Trilinear filtering disabled (not an OpenGL renderer)
DEBUG: Demuxer 'video': starting thread
DEBUG: Demuxer 'audio': starting thread
WARN: Device disconnected
ERROR: Demuxer 'video': stream disabled due to connection error
WARN: Demuxer 'audio': stream explicitly disabled by the device
DEBUG: quit...
DEBUG: Controller stopped
INFO: shutdown socket
INFO: shutdown socket
INFO: shutdown socket
Process 23503 exited with status = 2 (0x00000002) 
```

**Comment by metayan** [comment]
When building:
```
./server/build_without_gradle.sh && meson setup x --buildtype=debug \
  -Dprebuilt_server=build_manual/scrcpy-server && ninja -Cx
```
```
Platform: android-35                                                                                                                     
Build-tools: 35.0.0                                                                                                                      
Build dir: /Users/Shared/src/scm/android/scrcpy/build_manual                                                                             
Generating java from aidl...                            
Compiling java sources...              
Dexing...                                                                                                                                
Server generated in /Users/Shared/src/scm/android/scrcpy/build_manual/scrcpy-server
The Meson build system                                                                                                                   
Version: 1.4.1                                                                                                                           
Source dir: /Users/Shared/src/scm/android/scrcpy                                                                                         
Build dir: /Users/Shared/src/scm/android/scrcpy/x                                                                                        
Build type: native build                                          
Project name: scrcpy                                                                                                                     
Project version: 3.0                                                                                                                     
C compiler for the host machine: cc (clang 18.1.8 "Homebrew clang version 18.1.8")
C linker for the host machine: cc ld64 609.8
Host machine cpu family: x86_64
Host machine cpu: x86_64
Found pkg-config: YES (/usr/local/bin/pkg-config) 2.3.0
Run-time dependency libavformat found: YES 60.3.100
Run-time dependency libavcodec found: YES 60.3.100
Run-time dependency libavutil found: YES 58.2.100
Run-time dependency libswresample found: YES 4.10.100
Run-time dependency sdl2 found: YES 2.30.8
Run-time dependency libusb-1.0 found: YES 1.0.27
Checking for function "strdup" : YES 
Checking for function "asprintf" : YES 
Checking for function "vasprintf" : YES 
Checking for function "nrand48" : YES 
Checking for function "jrand48" : YES 
Checking for function "reallocarray" : NO 
Header "sys/socket.h" has symbol "SOCK_CLOEXEC" : NO 
Configuring config.h using configuration
Build targets in project: 13

scrcpy 3.0

  User defined options
    buildtype      : debug
    prebuilt_server: build_manual/scrcpy-server

Found ninja-1.12.1 at /usr/local/bin/ninja
ninja: Entering directory `x'
[129/129] Linking target app/scrcpy
```

**Comment by rom1v** [comment]
> OK, so your shutdown() does not interrupt accept(). Weird. 🤔

I guess we have to close() the socket in addition (like on Windows) in `net_interrupt()`.

Could you try this dirty hack:

```diff
diff --git app/src/util/intr.c app/src/util/intr.c
index 22bd121a7..48171237e 100644
--- app/src/util/intr.c
+++ app/src/util/intr.c
@@ -63,6 +63,7 @@ sc_intr_interrupt(struct sc_intr *intr) {
     if (intr->socket != SC_SOCKET_NONE) {
         LOGD("Interrupting socket");
         net_interrupt(intr->socket);
+        net_close(intr->socket); // hack
         intr->socket = SC_SOCKET_NONE;
     }
     if (intr->process != SC_PROCESS_NONE) {
```

**Comment by metayan** [comment]
[OKAAYY!](https://www.dailymotion.com/video/x2htyz0)

```lldb -- x/app/scrcpy --new-display```
```
(lldb) target create "x/app/scrcpy"
Current executable set to '/src/scm/android/scrcpy/x/app/scrcpy' (x86_64).
(lldb) settings set -- target.run-args  "--new-display"
(lldb) run
Process 23943 launched: '/src/scm/android/scrcpy/x/app/scrcpy' (x86_64)
scrcpy 3.0 <https://github.com/Genymobile/scrcpy>
INFO: ADB device found:
INFO:     -->   (usb)  xx                      device  G8441
DEBUG: Device serial: xx
DEBUG: Using SCRCPY_SERVER_PATH: x/server/scrcpy-server
x/server/scrcpy-server: 1 file pushed, 0 skipped. 229.3 MB/s (90877 bytes in 0.000s)
INFO: sc_server_connect_to()
INFO: sc_process_observer wait
INFO: accept()
[server] INFO: Device: [Sony] Sony G8441 (Android 9)
[server] ERROR: New virtual display is not supported before Android 10
INFO: sc_process_observer wait end
INFO: sc_server_on_terminated()
DEBUG: Interrupting socket
INFO: shutdown socket
INFO: accept() end
DEBUG: Server disconnected
DEBUG: Server terminated
WARN: Could not close server socket
INFO: sc_server_connect_to() returned false
INFO: sc_process_terminate(23957)
INFO: sc_process_wait(23957, true)
INFO: sc_process_observer_join()
INFO: sc_process_observer_destroy()
INFO: connection_failed
ERROR: Server connection failed
Process 23943 exited with status = 1 (0x00000001) 
```

**Comment by rom1v** [comment]
Please test branch [`macos_close_on_interrupt`](https://github.com/Genymobile/scrcpy/commits/macos_close_on_interrupt/).

**Comment by metayan** [comment]
Building generates lots of these warnings:
```
../app/src/util/net.h:33:10: warning: 'SC_SOCKET_NONE' macro redefined [-Wmacro-redefined]
   33 | # define SC_SOCKET_NONE NULL
      |          ^
../app/src/util/net.h:15:10: note: previous definition is here
   15 | # define SC_SOCKET_NONE -1
      |          ^
```

But it works fine now:
```x/app/scrcpy --new-display```
```
scrcpy 3.0 <https://github.com/Genymobile/scrcpy>
INFO: ADB device found:
INFO:     -->   (usb)  xx                      device  G8441
DEBUG: Device serial: xx
DEBUG: Using SCRCPY_SERVER_PATH: x/server/scrcpy-server
x/server/scrcpy-server: 1 file pushed, 0 skipped. 127.5 MB/s (90879 bytes in 0.001s)
[server] INFO: Device: [Sony] Sony G8441 (Android 9)
[server] ERROR: New virtual display is not supported before Android 10
DEBUG: Interrupting socket
DEBUG: Server disconnected
DEBUG: Server terminated
ERROR: Server connection failed
```

Also tried
```x/app/scrcpy --video-encoder=junk```
and plain
```x/app/scrcpy```
and they also work fine.

**Comment by rom1v** [comment]
Oops, please retry with this additional change:

```diff
diff --git app/src/util/net.h app/src/util/net.h
index d1e14df08..94789954a 100644
--- app/src/util/net.h
+++ app/src/util/net.h
@@ -12,7 +12,6 @@
 # define SC_RAW_SOCKET_NONE INVALID_SOCKET
 #else // not _WIN32
 # include <sys/socket.h>
-# define SC_SOCKET_NONE -1
   typedef int sc_raw_socket;
 # define SC_RAW_SOCKET_NONE -1
 #endif
```

**Comment by metayan** [comment]
Was already testing it. ;)
Builds and works perfectly now.

**Comment by rom1v** [comment]
Merged into `dev` :rocket:
